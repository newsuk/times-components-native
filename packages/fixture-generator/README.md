# FixtureGenerator

This package is meant to create a single testable location for our fixtures. Enabling us to generate fixture data using schema introspection and graphql gen that is typesafe and reliable.

The aim for this package is to give us a single location for integration tests, showcase and UI tests all to use.

Notes:
* Since Jest depends on javascript, this package needs to be transpiled before running any Jest tests. Make sure to re-transpile after making any changes on this package.

* If new types are added to TPA. Please delete `schema.json` and `types.ts` before running `yarn getSchema` followed by `yarn gentypes`. This is due to an issue with file permissions. `types.ts` is an autogenerated file and should not be changed manually.

