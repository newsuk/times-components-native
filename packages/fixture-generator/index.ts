import MockArticle from "./src/mock-article";
import MockUser from "./src/mock-user";
import MockImage from "./src/mock-image";
import MockAuthor from "./src/mock-author";
import MockEdition from "./src/mock-edition";
import {
  mockMagazineSection,
  mockPuzzleSection,
  mockStandardSection,
  getPuzzleSlices,
} from "./src/mock-section";
import mockEditionSlice, {
  mockDailyRegisterSlice,
  mockLeadersSlice,
  mockCommentLeadAndCartoonSlice,
  mockStandardSlice,
  mockInTodaysEditionSlice,
  mockLeadOneAndFourSlice,
  mockLeadOneAndOneSlice,
  mockLeadOneAndTwoSlice,
  mockLeadOneFullWidthSlice,
  mockLeadTwoNoPicAndTwoSlice,
  mockListTwoAndSixNoPicSlice,
  mockSecondaryFourSlice,
  mockSecondaryOneSlice,
  mockSecondaryOneAndColumnistSlice,
  mockSecondaryOneAndFourSlice,
  mockSecondaryTwoAndTwoSlice,
  mockSecondaryTwoNoPicAndTwoSlice,
  mockPuzzleSlice,
  mockLeadTwoNoPicAndTwoFrontSlice,
  mockLeadOneAndOneFrontSlice,
  mockLeadOneFullWidthFrontSlice,
} from "./src/mock-slice";
import MockTile from "./src/mock-tile";
import MockTopic from "./src/mock-topic";

export {
  mockDailyRegisterSlice,
  mockLeadersSlice,
  mockCommentLeadAndCartoonSlice,
  mockStandardSlice,
  mockInTodaysEditionSlice,
  mockLeadOneAndFourSlice,
  mockLeadOneAndOneSlice,
  mockLeadOneAndTwoSlice,
  mockLeadOneFullWidthSlice,
  mockLeadTwoNoPicAndTwoSlice,
  mockListTwoAndSixNoPicSlice,
  mockSecondaryFourSlice,
  mockSecondaryOneSlice,
  mockSecondaryOneAndColumnistSlice,
  mockSecondaryOneAndFourSlice,
  mockSecondaryTwoAndTwoSlice,
  mockSecondaryTwoNoPicAndTwoSlice,
  mockMagazineSection,
  mockPuzzleSection,
  getPuzzleSlices,
  mockStandardSection,
  MockUser,
  MockArticle,
  MockEdition,
  mockEditionSlice,
  MockImage,
  MockAuthor,
  MockTile,
  MockTopic,
  mockPuzzleSlice,
  mockLeadTwoNoPicAndTwoFrontSlice,
  mockLeadOneAndOneFrontSlice,
  mockLeadOneFullWidthFrontSlice,
};
