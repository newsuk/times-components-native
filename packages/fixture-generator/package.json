{
  "name": "@times-components-native/fixture-generator",
  "private": true,
  "version": "0.28.1",
  "description": "Creates fixture data for testing",
  "main": "src/index.ts",
  "types": "src/types.ts",
  "scripts": {
    "transpile": "yarn getSchema && yarn gentypes && tsc",
    "gentypes": "gql-gen --schema ./schema.json --template graphql-codegen-typescript-template --out ./src/types.ts",
    "getSchema": "GRAPHQL_ENDPOINT=https://api.thetimes.co.uk/graphql getSchema",
    "test": "jest ./src/__tests__/*",
    "lint": "yarn prettier:diff && yarn depcheck",
    "fmt": "prettier --write 'src/*.*'",
    "prettier:diff": "prettier --list-different 'src/*.*'",
    "depcheck": "depcheck --ignores='@babel/*,babel-*,depcheck,jest,prettier,@times-components-native/schema,graphql-codegen-typescript-template,@types/jest,graphql-code-generator,ts-jest,typescript,typings' --ignore-bin-package=false --skip-missing"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/newsuk/times-components-native.git"
  },
  "jest": {
    "transform": {
      "^.+\\.ts?$": "ts-jest"
    },
    "testRegex": "(/__tests__/.*|(\\.|/)(test|spec))\\.ts?$",
    "moduleFileExtensions": [
      "ts",
      "js"
    ]
  },
  "keywords": [
    "react",
    "native",
    "web",
    "fixture-generator",
    "component"
  ],
  "license": "BSD-3-Clause",
  "bugs": {
    "url": "https://github.com/newsuk/times-components/issues"
  },
  "homepage": "https://github.com/newsuk/times-components#readme",
  "devDependencies": {
    "@times-components-native/schema": "file:../schema"
  },
  "publishConfig": {
    "access": "public"
  }
}
