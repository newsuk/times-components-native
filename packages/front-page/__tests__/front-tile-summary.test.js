import ReactTestRenderer from "react-test-renderer";
import React from "react";
import FrontTileSummary from "../front-tile-summary";
import { MockTile } from "@times-components-native/fixture-generator";

jest.mock(
  "../front-article-summary-content",
  () => "FrontArticleSummaryContent",
);
jest.mock("@times-components-native/article-summary", () => ({
  ArticleSummaryHeadline: "ArticleSummaryHeadline",
  ArticleSummaryStrapline: "ArticleSummaryStrapline",
}));

jest.mock("@times-components-native/article-byline", () => "ArticleByline");
jest.mock("@times-components-native/front-page/front-page-byline", () => ({
  FrontPageByline: "FrontPageByline",
}));

jest.mock("react-native-image-zoom-viewer", () => "ImageZoomView");

const mockContainerHeight = 200;
jest.mock("@times-components-native/front-page/MeasureContainer", () => {
  const MockMeasureContainer = ({ render }) => {
    return render({ height: mockContainerHeight, width: 200 });
  };
  return { MeasureContainer: MockMeasureContainer };
});

jest.mock("@times-components-native/responsive", () => ({
  useResponsiveContext: () => ({
    orientation: undefined,
  }),
}));

const summaryContent = [
  {
    attributes: {},
    children: [
      {
        attributes: {
          value: "Test",
        },
        children: [],
        name: "text",
      },
    ],
    name: "paragraph",
  },
];

const bylines = [
  {
    byline: [
      {
        foo: "bar",
      },
    ],
  },
];

const props = {
  summary: summaryContent,
  tile: new MockTile().get(),
  whiteSpaceHeight: 40,
  bylines: bylines,
  containerStyle: { backgroundColor: "red" },
  headlineStyle: { backgroundColor: "blue" },
  straplineStyle: { backgroundColor: "green" },
  strapline: "Strapline Text",
  summaryStyle: { backgroundColor: "orange" },
  headlineMarginBottom: 10,
  straplineMarginTop: 10,
  straplineMarginBottom: 20,
  bylineMarginBottom: 15,
  summaryLineHeight: 20,
};

const simulateMeasurement = (
  renderer,
  { headlineHeight = 20, straplineHeight = 20, bylineHeight = 20 },
) => {
  // this is triggering the useEffect which simulates an orientation change
  ReactTestRenderer.act(() => {
    renderer.root
      .findByProps({
        testID: "headlineWrapper",
      })
      .props["onLayout"]({
        nativeEvent: { layout: { height: headlineHeight } },
      });
  });
  ReactTestRenderer.act(() => {
    renderer.root
      .findByProps({
        testID: "headlineWrapper",
      })
      .props["onLayout"]({
        nativeEvent: { layout: { height: headlineHeight } },
      });

    renderer.root
      .findByProps({
        testID: "straplineWrapper",
      })
      .props["onLayout"]({
        nativeEvent: { layout: { height: straplineHeight } },
      });

    renderer.root
      .findByProps({
        testID: "bylineWrapper",
      })
      .props["onLayout"]({ nativeEvent: { layout: { height: bylineHeight } } });
  });
};

describe("FrontTileSummary", () => {
  it("performs measurements before rendering front tile summary", () => {
    let renderer = ReactTestRenderer.create(<FrontTileSummary {...props} />);

    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it("shows front tile summary with headline only after measurement has been taken", () => {
    let renderer = ReactTestRenderer.create(<FrontTileSummary {...props} />);
    simulateMeasurement(renderer, {
      headlineHeight: mockContainerHeight,
      bylineHeight: 20,
      straplineHeight: 20,
    });

    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it("shows front tile summary with headline/strapline/byline/content after measurement has been taken", () => {
    let renderer = ReactTestRenderer.create(<FrontTileSummary {...props} />);
    simulateMeasurement(renderer, {
      headlineHeight: 20,
      bylineHeight: 20,
      straplineHeight: 20,
    });
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it("renders with a missing byline", () => {
    let renderer = ReactTestRenderer.create(
      <FrontTileSummary {...props} bylines={undefined} />,
    );

    simulateMeasurement(renderer, {
      headlineHeight: 20,
      bylineHeight: 0,
      straplineHeight: 20,
    });

    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it("renders with a missing strapline", () => {
    let renderer = ReactTestRenderer.create(
      <FrontTileSummary {...props} strapline={undefined} />,
    );

    simulateMeasurement(renderer, {
      headlineHeight: 20,
      bylineHeight: 20,
      straplineHeight: 0,
    });

    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it("renders with 2 columns", () => {
    let renderer = ReactTestRenderer.create(
      <FrontTileSummary {...props} columnCount={2} />,
    );

    simulateMeasurement(renderer, {
      headlineHeight: 20,
      bylineHeight: 20,
      straplineHeight: 20,
    });

    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it("renders byline with keyline", () => {
    let renderer = ReactTestRenderer.create(
      <FrontTileSummary {...props} showKeyline={true} />,
    );

    simulateMeasurement(renderer, {
      headlineHeight: 20,
      bylineHeight: 20,
      straplineHeight: 20,
    });

    expect(renderer.toJSON()).toMatchSnapshot();
  });
});
