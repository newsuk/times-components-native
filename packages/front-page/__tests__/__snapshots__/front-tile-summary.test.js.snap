// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FrontTileSummary renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "red",
      },
      Object {
        "flex": 1,
      },
    ]
  }
>
  <ArticleSummaryHeadline
    headline="This is tile headline"
    style={
      Object {
        "backgroundColor": "blue",
      }
    }
  />
  <ArticleSummaryStrapline
    strapline="Strapline Text"
    style={
      Object {
        "backgroundColor": "green",
      }
    }
  />
  <FrontPageByline
    byline={
      Array [
        Object {
          "byline": Array [
            Object {
              "foo": "bar",
            },
          ],
        },
      ]
    }
  />
  <FrontArticleSummaryContent
    bylines={
      Array [
        Object {
          "byline": Array [
            Object {
              "foo": "bar",
            },
          ],
        },
      ]
    }
    summary={
      Array [
        Object {
          "attributes": Object {},
          "children": Array [
            Object {
              "attributes": Object {
                "value": "Test",
              },
              "children": Array [],
              "name": "text",
            },
          ],
          "name": "paragraph",
        },
      ]
    }
    summaryStyle={
      Object {
        "backgroundColor": "orange",
      }
    }
    template="mainstandard"
  />
</View>
`;

exports[`FrontTileSummary renders with keyline 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "red",
      },
      Object {
        "flex": 1,
      },
    ]
  }
>
  <ArticleSummaryHeadline
    headline="This is tile headline"
    style={
      Object {
        "backgroundColor": "blue",
      }
    }
  />
  <ArticleSummaryStrapline
    strapline="Strapline Text"
    style={
      Object {
        "backgroundColor": "green",
      }
    }
  />
  <FrontPageByline
    byline={
      Array [
        Object {
          "byline": Array [
            Object {
              "foo": "bar",
            },
          ],
        },
      ]
    }
    showKeyline={true}
  />
  <FrontArticleSummaryContent
    bylines={
      Array [
        Object {
          "byline": Array [
            Object {
              "foo": "bar",
            },
          ],
        },
      ]
    }
    summary={
      Array [
        Object {
          "attributes": Object {},
          "children": Array [
            Object {
              "attributes": Object {
                "value": "Test",
              },
              "children": Array [],
              "name": "text",
            },
          ],
          "name": "paragraph",
        },
      ]
    }
    summaryStyle={
      Object {
        "backgroundColor": "orange",
      }
    }
    template="mainstandard"
  />
</View>
`;

exports[`FrontTileSummary renders with more than 1 columns 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "red",
      },
      Object {
        "flex": 1,
      },
    ]
  }
>
  <ArticleSummaryHeadline
    headline="This is tile headline"
    style={
      Object {
        "backgroundColor": "blue",
      }
    }
  />
  <ArticleSummaryStrapline
    strapline="Strapline Text"
    style={
      Object {
        "backgroundColor": "green",
      }
    }
  />
  <FrontArticleSummaryContent
    bylines={
      Array [
        Object {
          "byline": Array [
            Object {
              "foo": "bar",
            },
          ],
        },
      ]
    }
    columnCount={2}
    summary={
      Array [
        Object {
          "attributes": Object {},
          "children": Array [
            Object {
              "attributes": Object {
                "value": "Test",
              },
              "children": Array [],
              "name": "text",
            },
          ],
          "name": "paragraph",
        },
      ]
    }
    summaryStyle={
      Object {
        "backgroundColor": "orange",
      }
    }
    template="mainstandard"
  />
</View>
`;

exports[`FrontTileSummary renders without byline 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "red",
      },
      Object {
        "flex": 1,
      },
    ]
  }
>
  <ArticleSummaryHeadline
    headline="This is tile headline"
    style={
      Object {
        "backgroundColor": "blue",
      }
    }
  />
  <ArticleSummaryStrapline
    strapline="Strapline Text"
    style={
      Object {
        "backgroundColor": "green",
      }
    }
  />
  <FrontArticleSummaryContent
    summary={
      Array [
        Object {
          "attributes": Object {},
          "children": Array [
            Object {
              "attributes": Object {
                "value": "Test",
              },
              "children": Array [],
              "name": "text",
            },
          ],
          "name": "paragraph",
        },
      ]
    }
    summaryStyle={
      Object {
        "backgroundColor": "orange",
      }
    }
    template="mainstandard"
  />
</View>
`;

exports[`FrontTileSummary renders without strapline 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "red",
      },
      Object {
        "flex": 1,
      },
    ]
  }
>
  <ArticleSummaryHeadline
    headline="This is tile headline"
    style={
      Object {
        "backgroundColor": "blue",
      }
    }
  />
  <FrontPageByline
    byline={
      Array [
        Object {
          "byline": Array [
            Object {
              "foo": "bar",
            },
          ],
        },
      ]
    }
  />
  <FrontArticleSummaryContent
    bylines={
      Array [
        Object {
          "byline": Array [
            Object {
              "foo": "bar",
            },
          ],
        },
      ]
    }
    summary={
      Array [
        Object {
          "attributes": Object {},
          "children": Array [
            Object {
              "attributes": Object {
                "value": "Test",
              },
              "children": Array [],
              "name": "text",
            },
          ],
          "name": "paragraph",
        },
      ]
    }
    summaryStyle={
      Object {
        "backgroundColor": "orange",
      }
    }
    template="mainstandard"
  />
</View>
`;
